File: core/bmi.py
import math


def lbs_to_kg(lbs: float) -> float:
    return lbs * 0.453592


def kg_to_lbs(kg: float) -> float:
    return kg / 0.453592


def inches_to_cm(inches: float) -> float:
    return inches * 2.54


def cm_to_inches(cm: float) -> float:
    return cm / 2.54


def calc_standard_bmi(weight_kg: float, height_m: float) -> float:
    return weight_kg / (height_m ** 2)


def calc_new_bmi(weight_kg: float, height_m: float) -> float:
    return 1.3 * weight_kg / (height_m ** 2.5)


def calc_ponderal_index(weight_kg: float, height_m: float) -> float:
    return weight_kg / (height_m ** 3)


def calc_bsa(height_cm: float, weight_kg: float) -> float:
    return math.sqrt(height_cm * weight_kg / 3600)


def calc_whtr(waist_cm: float, height_cm: float) -> float:
    return waist_cm / height_cm


def _inches_over_5ft(height_cm: float) -> float:
    return max(0.0, height_cm / 2.54 - 60.0)


def calc_ibw_hamwi(height_cm: float, sex: str) -> float:
    x = _inches_over_5ft(height_cm)
    if sex == "Male":
        return 48.0 + 2.7 * x
    elif sex == "Female":
        return 45.5 + 2.2 * x
    else:
        return (48.0 + 2.7 * x + 45.5 + 2.2 * x) / 2


def calc_ibw_devine(height_cm: float, sex: str) -> float:
    x = _inches_over_5ft(height_cm)
    if sex == "Male":
        return 50.0 + 2.3 * x
    elif sex == "Female":
        return 45.5 + 2.3 * x
    else:
        return (50.0 + 2.3 * x + 45.5 + 2.3 * x) / 2


def calc_ibw_robinson(height_cm: float, sex: str) -> float:
    x = _inches_over_5ft(height_cm)
    if sex == "Male":
        return 52.0 + 1.9 * x
    elif sex == "Female":
        return 49.0 + 1.7 * x
    else:
        return (52.0 + 1.9 * x + 49.0 + 1.7 * x) / 2


def calc_ibw_miller(height_cm: float, sex: str) -> float:
    x = _inches_over_5ft(height_cm)
    if sex == "Male":
        return 56.2 + 1.41 * x
    elif sex == "Female":
        return 53.1 + 1.36 * x
    else:
        return (56.2 + 1.41 * x + 53.1 + 1.36 * x) / 2

File: mobile/lib/core/bmi.dart
import 'dart:math';

// ── Unit conversions ─────────────────────────────────────────────────────────

double lbsToKg(double lbs) => lbs * 0.453592;
double kgToLbs(double kg) => kg / 0.453592;
double inchesToCm(double inches) => inches * 2.54;
double cmToInches(double cm) => cm / 2.54;

// ── Indices ───────────────────────────────────────────────────────────────────

double calcStandardBmi(double weightKg, double heightM) =>
    weightKg / (heightM * heightM);

double calcNewBmi(double weightKg, double heightM) =>
    1.3 * weightKg / pow(heightM, 2.5);

double calcPonderalIndex(double weightKg, double heightM) =>
    weightKg / pow(heightM, 3);

double calcBsa(double heightCm, double weightKg) =>
    sqrt(heightCm * weightKg / 3600);

double calcWhtr(double waistCm, double heightCm) => waistCm / heightCm;

// ── Body fat ──────────────────────────────────────────────────────────────────

double? calcBfNavyMale(double heightCm, double waistCm, double neckCm) {
  final val = waistCm - neckCm;
  if (val <= 0) return null;
  return 495 / (1.0324 - 0.19077 * log(val) / ln10 + 0.15456 * log(heightCm) / ln10) - 450;
}

double? calcBfNavyFemale(
    double heightCm, double waistCm, double hipCm, double neckCm) {
  final val = waistCm + hipCm - neckCm;
  if (val <= 0) return null;
  return 495 / (1.29579 - 0.35004 * log(val) / ln10 + 0.22100 * log(heightCm) / ln10) - 450;
}

double calcBfDeurenberg(double bmi, double age, String sex) {
  final sexFactor = sex == 'Male' ? 1.0 : (sex == 'Female' ? 0.0 : 0.5);
  return 1.20 * bmi + 0.23 * age - 10.8 * sexFactor - 5.4;
}

// ── Ideal body weight ─────────────────────────────────────────────────────────

double _inchesOver5ft(double heightCm) =>
    max(0.0, heightCm / 2.54 - 60.0);

double calcIbwHamwi(double heightCm, String sex) {
  final x = _inchesOver5ft(heightCm);
  if (sex == 'Male') return 48.0 + 2.7 * x;
  if (sex == 'Female') return 45.5 + 2.2 * x;
  return ((48.0 + 2.7 * x) + (45.5 + 2.2 * x)) / 2;
}

double calcIbwDevine(double heightCm, String sex) {
  final x = _inchesOver5ft(heightCm);
  if (sex == 'Male') return 50.0 + 2.3 * x;
  if (sex == 'Female') return 45.5 + 2.3 * x;
  return ((50.0 + 2.3 * x) + (45.5 + 2.3 * x)) / 2;
}

double calcIbwRobinson(double heightCm, String sex) {
  final x = _inchesOver5ft(heightCm);
  if (sex == 'Male') return 52.0 + 1.9 * x;
  if (sex == 'Female') return 49.0 + 1.7 * x;
  return ((52.0 + 1.9 * x) + (49.0 + 1.7 * x)) / 2;
}

double calcIbwMiller(double heightCm, String sex) {
  final x = _inchesOver5ft(heightCm);
  if (sex == 'Male') return 56.2 + 1.41 * x;
  if (sex == 'Female') return 53.1 + 1.36 * x;
  return ((56.2 + 1.41 * x) + (53.1 + 1.36 * x)) / 2;
}
